File: leetcode/3400-3500/3479. 水果成篮 III.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *
from random import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float("inf")
MOD = int(1e9 + 7)


class SegmentTree:
    def __init__(self, nums: List[int]) -> None:
        self.n = len(nums)
        self.tree = [0] * (2 << ((self.n - 1).bit_length()))
        self.build(nums, 1, 0, self.n - 1)

    def maintain(self, o):
        self.tree[o] = max(self.tree[o * 2], self.tree[o * 2 + 1])

    def build(self, nums, o, l, r):
        if l == r:
            self.tree[o] = nums[l]
            return
        m = (l + r) // 2
        self.build(nums, o * 2, l, m)
        self.build(nums, o * 2 + 1, m + 1, r)
        self.maintain(o)

    def query(self, o, l, r, target):
        if self.tree[o] < target:
            return -1
        if l == r:
            if self.tree[o] >= target:
                self.tree[o] = -1
                return 1
            else:
                return -1

        m = (l + r) // 2
        t = self.query(o * 2, l, m, target)
        if t < 0:
            t = self.query(o * 2 + 1, m + 1, r, target)
        self.maintain(o)
        return t


class Solution:
    def numOfUnplacedFruits(self, fruits: List[int], baskets: List[int]) -> int:
        st = SegmentTree(baskets)
        ans = 0
        for x in fruits:
            i = st.query(1, 0, len(baskets) - 1, x)
            if i < 0:
                ans += 1
        return ans


class SegmentTree:
    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.tree = [-1] * (2 << ((self.n - 1).bit_length()))
        self.build(nums, 1, 0, self.n - 1)

    def build(self, nums, o, l, r):
        if l == r:
            self.tree[o] = nums[l]
            return
        m = (l + r) // 2
        self.build(nums, o * 2, l, m)
        self.build(nums, o * 2 + 1, m + 1, r)
        self.maintain(o)

    def maintain(self, o):
        self.tree[o] = max(self.tree[o * 2], self.tree[o * 2 + 1])

    def query(self, o, l, r, target):
        if self.tree[o] < target:
            return -1

        if l == r:
            self.tree[o] = -1
            return 0
        m = (l + r) // 2
        i = self.query(o * 2, l, m, target)
        if i < 0:
            i = self.query(o * 2 + 1, m + 1, r, target)
        self.maintain(o)
        return i


class Solution:
    def numOfUnplacedFruits(self, fruits: List[int], baskets: List[int]) -> int:
        st = SegmentTree(baskets)
        ans = 0
        n = len(baskets)
        for x in fruits:
            i = st.query(1, 0, n - 1, x)
            if i < 0:
                ans += 1
        return ans

File: leetcode/3400-3500/3479. 水果成篮 III.cpp
//
// Created by 123 on 25-7-24.
//
#include <bit>
#include <bits/stdc++.h>
#include <vector>
using namespace std;
using ll = long long;
using pii = pair<int, int>;
using pll = pair<ll, ll>;
#define For(i, n) for (int(i) = 0; (i) < (n); (i) += 1)
constexpr int MOD = int(1e9 + 7);
#define i128 __int128_t
#define ull unsigned long long
const ll MOD2 = 4611686018427387847;

class St {
public:
    vector<int> tree;
    St(int n) { tree.resize(2 << (bit_width((unsigned) n))); }

    void update(int o, int l, int r, int i, int val) {
        if (l == r) {
            tree[o] = val;
            return;
        }
        int m = l + (r - l) / 2;
        if (i <= m) {
            update(o * 2, l, m, i, val);
        } else {
            update(o * 2 + 1, m + 1, r, i, val);
        }
        tree[o] = max(tree[o * 2], tree[o * 2 + 1]);
    }

    int query(int o, int l, int r, int target) {
        if (tree[o] < target) {
            return -1;
        }
        if (l == r) {
            return l;
        }
        int m = l + (r - l) / 2;
        int res = query(o * 2, l, m, target);
        if (res < 0) {
            res = query(o * 2 + 1, m + 1, r, target);
        }
        return res;
    }
};


class Solution {
public:
    int numOfUnplacedFruits(vector<int> &fruits, vector<int> &baskets) {
        int n = fruits.size();
        St st(n);
        for (int i = 0; i < n; i++) {
            st.update(1, 0, n - 1, i, baskets[i]);
        }

        int ans = 0;
        for (int i = 0; i < n; i++) {
            int f = fruits[i];
            int j = st.query(1, 0, n - 1, f);
            if (j < 0) {
                ans++;
            } else {
                st.update(1, 0, n - 1, j, 0);
            }
        }
        return ans;
    }
};






class SegmentTree {
    vector<int> mx;

    void maintain(int o) { mx[o] = max(mx[o * 2], mx[o * 2 + 1]); }

    // 初始化线段树
    void build(const vector<int> &a, int o, int l, int r) {
        if (l == r) {
            mx[o] = a[l];
            return;
        }
        int m = (l + r) / 2;
        build(a, o * 2, l, m);
        build(a, o * 2 + 1, m + 1, r);
        maintain(o);
    }

public:
    SegmentTree(const vector<int> &a) {
        size_t n = a.size();
        mx.resize(2 << bit_width(n - 1));
        build(a, 1, 0, n - 1);
    }

    // 找区间内的第一个 >= x 的数，并更新为 -1，返回这个数的下标（没有则返回 -1）
    int findFirstAndUpdate(int o, int l, int r, int x) {
        if (mx[o] < x) {
            // 区间没有 >= x 的数
            return -1;
        }
        if (l == r) {
            mx[o] = -1; // 更新为 -1，表示不能放水果
            return l;
        }
        int m = (l + r) / 2;
        int i = findFirstAndUpdate(o * 2, l, m, x); // 先递归左子树
        if (i < 0) {
            // 左子树没找到
            i = findFirstAndUpdate(o * 2 + 1, m + 1, r, x); // 再递归右子树
        }
        maintain(o);
        return i;
    }
};

class Solution {
public:
    int numOfUnplacedFruits(vector<int> &fruits, vector<int> &baskets) {
        SegmentTree t(baskets);
        int n = baskets.size(), ans = 0;
        for (int x: fruits) {
            if (t.findFirstAndUpdate(1, 0, n - 1, x) < 0) {
                ans++;
            }
        }
        return ans;
    }
};
